docs/source/user/settings.rst
Available settings
------------------

Admin related settings
======================

These settings control details of the administration interface of immunity-radius.

.. note::

    The values of overridden settings fields do not change even when
    the global defaults are changed.

``IMMUNITY
_RADIUS_EDITABLE_ACCOUNTING``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Whether ``radacct`` entries are editable from the django admin or not.

``IMMUNITY
_RADIUS_EDITABLE_POSTAUTH``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Whether ``postauth`` logs are editable from the django admin or not.

``IMMUNITY
_RADIUS_GROUPCHECK_ADMIN``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Direct editing of group checks items is disabled by default because
these can be edited through inline items in the Radius Group
admin (Freeradius > Groups).

*This is done with the aim of simplifying the admin interface and avoid
overwhelming users with too many options*.

If for some reason you need to enable direct editing of group checks
you can do so by setting this to ``True``.

``IMMUNITY
_RADIUS_GROUPREPLY_ADMIN``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Direct editing of group reply items is disabled by default because
these can be edited through inline items in the Radius Group
admin (Freeradius > Groups).

*This is done with the aim of simplifying the admin interface and avoid
overwhelming users with too many options*.

If for some reason you need to enable direct editing of group replies
you can do so by setting this to ``True``.

``IMMUNITY
_RADIUS_USERGROUP_ADMIN``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Direct editing of user group items (``radusergroup``) is disabled by default
because these can be edited through inline items in the User
admin (Users and Organizations > Users).

*This is done with the aim of simplifying the admin interface and avoid
overwhelming users with too many options*.

If for some reason you need to enable direct editing of user group items
you can do so by setting this to ``True``.

``IMMUNITY
_RADIUS_USER_ADMIN_RADIUSTOKEN_INLINE``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

The functionality of editing a user's ``RadiusToken`` directly
through an inline from the user admin page is disabled by default.

*This is done with the aim of simplifying the admin interface and avoid
overwhelming users with too many options*.

If for some reason you need to enable editing user's ``RadiusToken``
from the user admin page, you can do so by setting this to ``True``.

Model related settings
======================

These settings control details of the immunity-radius model classes.

``IMMUNITY
_RADIUS_DEFAULT_SECRET_FORMAT``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``NT-Password``

The default encryption format for storing radius check values.

``IMMUNITY
_RADIUS_DISABLED_SECRET_FORMATS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``[]``

A list of disabled encryption formats, by default all formats are
enabled in order to keep backward compatibility with legacy systems.

``IMMUNITY
_RADIUS_BATCH_DEFAULT_PASSWORD_LENGTH``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``8``

The default password length of the auto generated passwords while
batch addition of users from the csv.

``IMMUNITY
_RADIUS_BATCH_DELETE_EXPIRED``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``18``

It is the number of months after which the expired users are deleted.

``IMMUNITY
_RADIUS_BATCH_PDF_TEMPLATE``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

It is the template used to generate the pdf when users are being generated using the batch add users feature using the prefix.

The value should be the absolute path to the template of the pdf.

``IMMUNITY
_RADIUS_EXTRA_NAS_TYPES``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``tuple()``

This setting can be used to add custom NAS types that can be used from the
admin interface when managing NAS instances.

For example, you want a custom NAS type called ``cisco``, you would add
the following to your project ``settings.py``:

.. code-block:: python

    IMMUNITY
_RADIUS_EXTRA_NAS_TYPES = (
        ('cisco', 'Cisco Router'),
    )

.. _immunity_radius_freeradius_allowed_hosts:

``IMMUNITY
_RADIUS_FREERADIUS_ALLOWED_HOSTS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``[]``

List of host IP addresses or subnets allowed to consume the freeradius
API endpoints (Authorize, Accounting and Postauth), i.e the value
of this option should be the IP address of your freeradius
instance. Example: If your freeradius instance is running on
the same host machine as Immunity, the value should be ``127.0.0.1``.
Similarly, if your freeradius instance is on a different host in
the private network, the value should be the private IP of freeradius
host like ``192.0.2.50``. If your freeradius is on a public network,
please use the public IP of your freeradius instance.

You can use subnets when freeradius is hosted on a variable IP, eg:

- ``198.168.0.0/24`` to allow the entire LAN.
- ``0.0.0.0/0`` to allow any address (useful for development / testing).

This value can be overridden per organization in the organization
change page. You can skip setting this option if you intend to set
it from organization change page for each organization.

.. image:: /images/freeradius_allowed_hosts.png
   :alt: Organization change page freeradius settings

.. code-block:: python

    IMMUNITY
_RADIUS_FREERADIUS_ALLOWED_HOSTS = ['127.0.0.1', '192.0.2.10', '192.168.0.0/24']

If this option and organization change page option are both
empty, then all freeradius API requests for the organization
will return ``403``.

.. _coa_enabled_setting:

``IMMUNITY
_RADIUS_COA_ENABLED``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False```

If set to ``True``, immunity-radius will update the NAS with the
user's current RADIUS attributes whenever the ``RadiusGroup`` of
user is changed. This allow enforcing of rate limits on active
RADIUS sessions without requiring users to re-authenticate. For
more details, :ref:`read the dedicated section for configuring
immunity-radius and NAS for using CoA <change_of_authorization>`.

This can be overridden for each organization separately
via the organization radius settings section of the admin interface.

.. image:: /images/organization_coa_enabled.png
   :alt: CoA enabled

```RADCLIENT_ATTRIBUTE_DICTIONARIES```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+----------+
| **type**:    | ``list`` |
+--------------+----------+
| **default**: | ``[]``   |
+--------------+----------+

List of absolute file paths of additional RADIUS dictionaries used
for RADIUS attribute mapping.

.. note::

    A `default dictionary <https://github.com/edge-servers/immunity-radius/blob/master/immunity_radius/radclient/dictionary>`_
    is shipped with immunity-radius. Any dictionary added using this setting
    will be used alongside the default dictionary.

``IMMUNITY
_RADIUS_MAX_CSV_FILE_SIZE``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+----------------------------+
| **type**:    | ``int``                    |
+--------------+----------------------------+
| **default**: |  `5 * 1024 * 1024` (5 MB)  |
+--------------+----------------------------+

This setting can be used to set the maximum size limit for firmware images, eg:

.. code-block:: python

    IMMUNITY
_RADIUS_MAX_CSV_FILE_SIZE = 10 * 1024 * 1024  # 10MB

.. note::

    The numeric value represents the size of files in bytes.
    Setting this to ``None`` will mean there's no max size.

``IMMUNITY
_RADIUS_PRIVATE_STORAGE_INSTANCE``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+--------------+-------------------------------------------------------------------------------------+
| **type**:    | ``str``                                                                             |
+--------------+-------------------------------------------------------------------------------------+
| **default**: |  ``immunity_radius.private_storage.storage.private_file_system_storage``            |
+--------------+-------------------------------------------------------------------------------------+

Dotted path to an instance of any one of the storage classes in
`private_storage <https://github.com/edoburu/django-private-storage#django-private-storage>`_.
This instance is used for storing csv files of batch imports of users.

By default, an instance of ``private_storage.storage.files.PrivateFileSystemStorage``
is used.

.. _immunity_radius_called_station_ids:

``IMMUNITY
_RADIUS_CALLED_STATION_IDS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``{}``

This setting allows to specify the parameters to connect to the different
OpenVPN management interfaces available for an organization. This setting is used by the
:ref:`convert_called_station_id <convert_called_station_id>` command.

It should contain configuration in following format:

.. code-block:: python

    IMMUNITY
_RADIUS_CALLED_STATION_IDS = {
        # UUID of the organization for which settings are being specified
        # In this example 'default'
        '<organization_uuid>': {
            'openvpn_config': [
                {
                    # Host address of OpenVPN management
                    'host': '<host>',
                    # Port of OpenVPN management interface. Defaults to 7505 (integer)
                    'port': 7506,
                    # Password of OpenVPN management interface (optional)
                    'password': '<management_interface_password>',
                }
            ],
            # List of CALLED STATION IDs that has to be converted,
            # These look like: 00:27:22:F3:FA:F1:gw1.immunity.org
            'unconverted_ids': ['<called_station_id>'],
        }
    }

``IMMUNITY
_RADIUS_CONVERT_CALLED_STATION_ON_CREATE``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

If set to ``True``, "Called Station ID" of a RADIUS session will be
converted (as per configuration defined in :ref:`IMMUNITY
_RADIUS_CALLED_STATION_IDS <immunity_radius_called_station_ids>`)
just after the RADIUS session is created.

.. _immunity_radius_openvpn_datetime_format:

``IMMUNITY
_RADIUS_OPENVPN_DATETIME_FORMAT``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``u'%a %b %d %H:%M:%S %Y'``

Specifies the datetime format of OpenVPN management status parser used by the
:ref:`convert_called_station_id <convert_called_station_id>`
command.

``IMMUNITY
_RADIUS_UNVERIFY_INACTIVE_USERS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``0`` (disabled)

Number of days from user's ``last_login`` after which the
user will be flagged as *unverified*.

When set to ``0``, the feature would be disabled and the user will
not be flagged as *unverified*.

``IMMUNITY
_RADIUS_DELETE_INACTIVE_USERS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``0`` (disabled)

Number of days from user's ``last_login`` after which the
user will be deleted.

When set to ``0``, the feature would be disabled and the user will
not be deleted.


API and user token related settings
===================================

These settings control details related to the API and the radius user token.

``IMMUNITY
_RADIUS_API_URLCONF``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``None``

Changes the urlconf option of django urls to point the RADIUS API
urls to another installed module, example, ``myapp.urls``
(useful when you have a seperate API instance.)

``IMMUNITY
_RADIUS_API_BASEURL``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``/`` (points to same server)

If you have a seperate instance of immunity-radius API on a
different domain, you can use this option to change the base of the image
download URL, this will enable you to point to your API server's domain,
example value: ``https://myradius.myapp.com``.

.. _immunity_radius_api:

``IMMUNITY
_RADIUS_API``
~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``True``

Indicates whether the REST API of immunity-radius is enabled or not.

``IMMUNITY
_RADIUS_DISPOSABLE_RADIUS_USER_TOKEN``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``True``

Radius user tokens are used for authorizing users.

When this setting is ``True`` radius user tokens are deleted right after a successful
authorization is performed. This reduces the possibility of attackers reusing
the access tokens and posing as other users if they manage to intercept it somehow.

.. _immunity_radius_api_authorize_reject:

``IMMUNITY
_RADIUS_API_AUTHORIZE_REJECT``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Indicates wether the :ref:`Authorize API view <authorize>` will return
``{"control:Auth-Type": "Reject"}`` or not.

Rejecting an authorization request explicitly will prevent freeradius from
attempting to perform authorization with other mechanisms (eg: radius checks, LDAP, etc.).

When set to ``False``, if an authorization request fails, the API will respond with
``None``, which will allow freeradius to keep attempting to authorize the request
with other freeradius modules.

Set this to ``True`` if you are performing authorization exclusively through the REST API.

``IMMUNITY
_RADIUS_API_ACCOUNTING_AUTO_GROUP``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``True``

When this setting is enabled, every accounting instance saved from the API will have
its ``groupname`` attribute automatically filled in.
The value filled in will be the ``groupname`` of the ``RadiusUserGroup`` of the highest
priority among the RadiusUserGroups related to the user with the ``username`` as in the
accounting instance.
In the event there is no user in the database corresponding to the ``username`` in the
accounting instance, the failure will be logged with ``warning`` level but the accounting
will be saved as usual.

.. _immunity_radius_allowed_mobile_prefixes:

``IMMUNITY
_RADIUS_ALLOWED_MOBILE_PREFIXES``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``[]``

This setting is used to specify a list of international mobile prefixes which should
be allowed to register into the system via the :ref:`user registration API <user_registration>`.

That is, only users with phone numbers using the specified international prefixes will
be allowed to register.

Leaving this unset or setting it to an empty list (``[]``) will effectively allow
any international mobile prefix to register (which is the default setting).

For example:

.. code-block:: python

    IMMUNITY
_RADIUS_ALLOWED_MOBILE_PREFIXES = ['+44', '+237']

Using the setting above will only allow phone numbers from the UK (``+44``)
or Cameroon (``+237``).

.. note::

    This setting is applicable only for organizations
    which have :ref:`enabled the SMS verification option
    <immunity_radius_sms_verification_enabled>`.

``IMMUNITY
_RADIUS_ALLOW_FIXED_LINE_OR_MOBILE``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**: ``False``

Immunity RADIUS only allow using mobile phone numbers for user registration.
This can cause issues in regions where fixed line and mobile phone numbers
 uses the same pattern (e.g. USA). Setting the value to ``True``
would make phone number type checking less strict.

.. _immunity_radius_optional_registration_fields:

``IMMUNITY
_RADIUS_OPTIONAL_REGISTRATION_FIELDS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Default**:

.. code-block:: python

    {
        'first_name': 'disabled',
        'last_name': 'disabled',
        'birth_date': 'disabled',
        'location': 'disabled',
    }

This global setting is used to specify if the optional user fields
(``first_name``, ``last_name``, ``location`` and ``birth_date``)

CHANGES.rst
Change log
==========

Version 1.1.0 [Unreleased]
--------------------------

Unreleased (Work in progress)

Version 1.0.2 [2022-12-05]
--------------------------

Bugfixes
~~~~~~~~

- Made private storage backend configurable
- Updated API views to use  ``filterset_class`` instead of ``filter_class``
  (required by ``django-filter==22.1``)
- Fixed organization cache bug in SAML ACS view: A forceful update of
  the user's organization cache is done before performing post-login
  operations to avoid issues occurring due to outdated cache.
- Added missing Furlan translation for sesame link validity
- Use storage backend method for deleting ``RadiusBatch.csvfile``:
  The previous implementation used the "os" module for deleting resisdual
  csv files. This causes issues when the project uses a file storage backend
  other than based on file system.
- Added error handling in RadiusBatch admin change view: Accessing admin change
  view of a non-existent RadiusBatch object resulted in Server Error 500
  because the ``DoesNotExist`` conditioned was not handled.
- Load image using ``static()`` in RegisteredUserInline.get_is_verified
- Use ``path`` URL kwarg in "serve_private_file" URL pattern
- Honor DISPOSABLE_RADIUS_USER_TOKEN in accounting stop API view:
  The accounting stop REST API operation was not taking into
  account the IMMUNITY
_RADIUS_DISPOSABLE_RADIUS_USER_TOKEN
  setting when disabling the auth capability of the radius token.

Version 1.0.1 [2022-05-10]
--------------------------

Bugfixes
~~~~~~~~

- Fixed a bug in the organization radius settings form which was causing it
  to not display some default values correctly
- Fixed a bug in allowed mobile prefix implementation:
  the implementation was joining the globally allowed prefixes
  and the prefixes allowed at org level, with the result
  that disabling a prefix at org level was not possible
- Called-station-ID command: log with warning instead of ``warn`` or ``error``:
  - warn > warning (warn is deprecated)
  - use warning instead of errors for more temporary connection issues cases

Version 1.0.0 [2022-04-18]
--------------------------

Features
~~~~~~~~

- Allowed to login via API with email or phone number
- Allowed freeradius authorize with email or phone number
- Allowed the usage of subnets in `IMMUNITY
_RADIUS_FREERADIUS_ALLOWED_HOSTS
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-freeradius-allowed-hosts>`_
- Made the fields containing personal data of users which are exposed in the registration API
  configurable (allowed, mandatory, disabled) via the
  `IMMUNITY
_RADIUS_OPTIONAL_REGISTRATION_FIELDS setting or the admin interface
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-optional-registration-fields>`_
- Allow to disable registration API via the
  `IMMUNITY
_RADIUS_REGISTRATION_API_ENABLED setting
  or the admin interface
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-registration-api-enabled>`_
- Added `throttling of API requests
  <https://immunity-radius.readthedocs.io/en/latest/user/api.html#api-throttling>`_
- Added `IMMUNITY
_RADIUS_API_BASEURL setting
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-api-baseurl>`_
- Add identity verification feature, configurable via the
  `IMMUNITY
_RADIUS_NEEDS_IDENTITY_VERIFICATION or via admin interface
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-needs-identity-verification>`_
- Added utilities for implementing
  `new registration and identity verification methods
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#adding-support-for-more-registration-verification-methods>`_
- Added `captive portal mock views
  <https://immunity-radius.readthedocs.io/en/latest/developer/captive_portal_mock.html>`_
  to ease development and debugging
- Add possibility to filter users by registration method in the admin interface
- Added SAML registration method to implement `captive portal authentication
  via Single Sign On (SSO) <https://immunity-radius.readthedocs.io/en/latest/user/saml.html>`_
- Added management command and celery task to
  `delete unverified users
  <https://immunity-radius.readthedocs.io/en/latest/user/management_commands.html#delete-unverified-users>`_
- Added translations of user facing API responses in Italian, German, Slovenian and Furlan
- Added `Convert RADIUS accounting CALLED-STATION-ID feature
  <https://immunity-radius.readthedocs.io/en/latest/user/management_commands.html#convert-called-station-id>`_,
  celery task and management command,
  with the possibility of triggering it on accounting creation
  (see `IMMUNITY
_RADIUS_CONVERT_CALLED_STATION_ON_CREATE
  <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-convert-called-station-on-create>`_)
- Added an `equivalent of the FreeRADIUS sqlcounter feature to the REST API
  <https://immunity-radius.readthedocs.io/en/latest/user/enforcing_limits.html#how-limits-are-enforced-counters>`_
- Added emission of django signal to FreeRADIUS accounting view:
  `radius_accounting_success
  <https://immunity-radius.readthedocs.io/en/latest/developer/signals.html#radius-accounting-success>`_
- Added possibility to send email to the user an they start
  a new radius accounting session
- Added organization level settings and related admin interface functionality
  to enable/disable SAML and social login:

  - `IMMUNITY
_RADIUS_SAML_REGISTRATION_ENABLED
    <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-saml-registration-enabled>`_
  - `IMMUNITY
_RADIUS_SOCIAL_REGISTRATION_ENABLED
    <https://immunity-radius.readthedocs.io/en/latest/user/settings.html#immunity-radius-social-registration-enabled>`_

- Added setting to avoid updating username from SAML:
  `IMMUNITY
_RADIUS_SAML_UPDATES_PRE_EXISTING_USERNAME

immunity_radius/settings.py
import hashlib
import logging
import os
from uuid import UUID

from django.conf import settings
from django.core.exceptions import ImproperlyConfigured
from django.utils.module_loading import import_string
from django.utils.translation import gettext_lazy as _

# 'pre_django_setup' is supposed to be a logger
# that can work before registered Apps are
# ready in django.setup() process.
logger = logging.getLogger('pre_django_setup')
DEBUG = settings.DEBUG


def get_settings_value(option, default):
    if hasattr(settings, f'DJANGO_FREERADIUS_{option}'):  # pragma: no cover
        logger.warning(
            f'DJANGO_FREERADIUS_{option} setting is deprecated. It will be '
            f'removed in the future, please use IMMUNITY
_RADIUS_{option} instead.'
        )
        return getattr(settings, f'DJANGO_FREERADIUS_{option}')
    return getattr(settings, f'IMMUNITY
_RADIUS_{option}', default)


def get_default_password_reset_url(urls):
    # default is kept for backward compatibility,
    # will be removed in future versions
    return urls.get('default') or urls.get('__all__')


RADIUS_API = get_settings_value('API', True)
RADIUS_API_BASEURL = get_settings_value('API_BASEURL', '/')
RADIUS_API_URLCONF = get_settings_value('API_URLCONF', None)
EDITABLE_ACCOUNTING = get_settings_value('EDITABLE_ACCOUNTING', False)
EDITABLE_POSTAUTH = get_settings_value('EDITABLE_POSTAUTH', False)
GROUPCHECK_ADMIN = get_settings_value('GROUPCHECK_ADMIN', False)
GROUPREPLY_ADMIN = get_settings_value('GROUPREPLY_ADMIN', False)
USERGROUP_ADMIN = get_settings_value('USERGROUP_ADMIN', False)
USER_ADMIN_RADIUSTOKEN_INLINE = get_settings_value(
    'USER_ADMIN_RADIUSTOKEN_INLINE', False
)
DEFAULT_SECRET_FORMAT = get_settings_value('DEFAULT_SECRET_FORMAT', 'NT-Password')
DISABLED_SECRET_FORMATS = get_settings_value('DISABLED_SECRET_FORMATS', [])
BATCH_DEFAULT_PASSWORD_LENGTH = get_settings_value('BATCH_DEFAULT_PASSWORD_LENGTH', 8)
BATCH_DELETE_EXPIRED = get_settings_value('BATCH_DELETE_EXPIRED', 18)
BATCH_MAIL_SUBJECT = get_settings_value('BATCH_MAIL_SUBJECT', 'Credentials')
BATCH_MAIL_SENDER = get_settings_value('BATCH_MAIL_SENDER', settings.DEFAULT_FROM_EMAIL)
API_AUTHORIZE_REJECT = get_settings_value('API_AUTHORIZE_REJECT', False)
SOCIAL_REGISTRATION_CONFIGURED = 'allauth.socialaccount' in getattr(
    settings, 'INSTALLED_APPS', []
)
SOCIAL_REGISTRATION_ENABLED = get_settings_value('SOCIAL_REGISTRATION_ENABLED', False)
SAML_REGISTRATION_CONFIGURED = 'djangosaml2' in getattr(settings, 'INSTALLED_APPS', [])
SAML_REGISTRATION_ENABLED = get_settings_value('SAML_REGISTRATION_ENABLED', False)
MAC_ADDR_ROAMING_ENABLED = get_settings_value('MAC_ADDR_ROAMING_ENABLED', False)
SAML_REGISTRATION_METHOD_LABEL = get_settings_value(
    'SAML_REGISTRATION_METHOD_LABEL', _('Single Sign-On (SAML)')
)
SAML_IS_VERIFIED = get_settings_value('SAML_IS_VERIFIED', False)
SAML_UPDATES_PRE_EXISTING_USERNAME = get_settings_value(
    'SAML_UPDATES_PRE_EXISTING_USERNAME', False
)
UNVERIFY_INACTIVE_USERS = get_settings_value('UNVERIFY_INACTIVE_USERS', 0)
DELETE_INACTIVE_USERS = get_settings_value('DELETE_INACTIVE_USERS', 0)
DISPOSABLE_RADIUS_USER_TOKEN = get_settings_value('DISPOSABLE_RADIUS_USER_TOKEN', True)
API_ACCOUNTING_AUTO_GROUP = get_settings_value('API_ACCOUNTING_AUTO_GROUP', True)
FREERADIUS_ALLOWED_HOSTS = get_settings_value('FREERADIUS_ALLOWED_HOSTS', [])
EXTRA_NAS_TYPES = get_settings_value('EXTRA_NAS_TYPES', tuple())
MAX_CSV_FILE_SIZE = get_settings_value('MAX_FILE_SIZE', 5 * 1024 * 1024)
BATCH_PDF_TEMPLATE = get_settings_value(
    'BATCH_PDF_TEMPLATE',
    os.path.join(
        os.path.dirname(__file__), 'templates/immunity-radius/prefix_pdf.html'
    ),
)
BATCH_MAIL_MESSAGE = get_settings_value(
    'BATCH_MAIL_MESSAGE', 'username: {}, password: {}'
)
PASSWORD_RESET_URLS = {
    # fallback in case the specific org page is not defined
    '__all__': 'https://{site}/{organization}/password/reset/confirm/{uid}/{token}',
    # use the uuid because the slug can change
}
PASSWORD_RESET_URLS.update(get_settings_value('PASSWORD_RESET_URLS', {}))
DEFAULT_PASSWORD_RESET_URL = get_default_password_reset_url(PASSWORD_RESET_URLS)
SMS_VERIFICATION_ENABLED = get_settings_value('SMS_VERIFICATION_ENABLED', False)
COA_ENABLED = get_settings_value('COA_ENABLED', False)
# SMS_TOKEN_DEFAULT_VALIDITY time is in minutes
SMS_TOKEN_DEFAULT_VALIDITY = get_settings_value('SMS_TOKEN_DEFAULT_VALIDITY', 30)
SMS_TOKEN_LENGTH = get_settings_value('SMS_TOKEN_LENGTH', 6)
SMS_TOKEN_HASH_ALGORITHM = get_settings_value('SMS_TOKEN_HASH_ALGORITHM', 'sha256')
SMS_TOKEN_MAX_ATTEMPTS = get_settings_value('SMS_TOKEN_MAX_ATTEMPTS', 5)
SMS_TOKEN_MAX_USER_DAILY = get_settings_value('SMS_TOKEN_MAX_USER_DAILY', 5)
SMS_COOLDOWN = get_settings_value('SMS_COOLDOWN', 30)
# default is high because immunity-wifi-login-pages results always as 1 IP
SMS_TOKEN_MAX_IP_DAILY = get_settings_value('SMS_TOKEN_MAX_IP_DAILY', 999)
ALLOWED_MOBILE_PREFIXES = get_settings_value('ALLOWED_MOBILE_PREFIXES', [])
ALLOW_FIXED_LINE_OR_MOBILE = get_settings_value('ALLOW_FIXED_LINE_OR_MOBILE', False)
REGISTRATION_API_ENABLED = get_settings_value('REGISTRATION_API_ENABLED', True)
NEEDS_IDENTITY_VERIFICATION = get_settings_value('NEEDS_IDENTITY_VERIFICATION', False)
SMS_MESSAGE_TEMPLATE = get_settings_value(
    'SMS_MESSAGE_TEMPLATE', '{organization} verification code: {code}'
)
OPTIONAL_REGISTRATION_FIELDS = get_settings_value(
    'OPTIONAL_REGISTRATION_FIELDS',
    {
        'first_name': 'disabled',
        'last_name': 'disabled',
        'birth_date': 'disabled',
        'location': 'disabled',
    },
)

try:  # pragma: no cover
    assert PASSWORD_RESET_URLS
    for key, value in PASSWORD_RESET_URLS.items():
        if key != '__all__' and key != 'default':
            try:
                UUID(key)
            except ValueError:
                raise AssertionError(f'{key} is not a valid UUID')
        assert all(['{organization}' in value, '{uid}' in value, '{token}' in value]), (
            f'{value} must contain ' + '{organization}, {uid} and {token}'  # noqa
        )
except AssertionError as error:  # pragma: no cover
    raise ImproperlyConfigured(
        f'IMMUNITY
_RADIUS_PASSWORD_RESET_URLS is invalid: {error}'
    )

try:
    SMS_TOKEN_HASH_ALGORITHM = getattr(hashlib, SMS_TOKEN_HASH_ALGORITHM)
except ImportError as error:  # pragma: no cover
    raise ImproperlyConfigured(
        f'IMMUNITY
_RADIUS_SMS_TOKEN_HASH_ALGORITHM is invalid: {error}'
    )

try:
    assert int(SMS_TOKEN_LENGTH) <= 8 and int(SMS_TOKEN_LENGTH) >= 4
except AssertionError:  # pragma: no cover
    raise ImproperlyConfigured(
        'IMMUNITY
_RADIUS_SMS_TOKEN_LENGTH must be a number between 4 and 8: '
        'lower would not be safe and higher would not be practical from '
        'a ux perspective.'
    )

# Path of urls that need to be refered in migrations files.
CSV_URL_PATH = 'api/v1/radius/organization/'
try:
    PRIVATE_STORAGE_INSTANCE = import_string(
        get_settings_value(
            'PRIVATE_STORAGE_INSTANCE',
            'immunity_radius.private_storage.storage.private_file_system_storage',
        )
    )
except ImportError:
    raise ImproperlyConfigured('Failed to import PRIVATE_STORAGE_INSTANCE')

CALLED_STATION_IDS = get_settings_value('CALLED_STATION_IDS', {})

for organization in CALLED_STATION_IDS.keys():  # pragma: no cover
    try:
        UUID(organization)
    except ValueError:
        logger.warning(
            'Organization slug in CALLED_STATION_IDS setting is deprecated. '
            'It will be removed in future, please replace '
            f'organization slug: {organization} with its id.'
        )
CONVERT_CALLED_STATION_ON_CREATE = get_settings_value(
    'CONVERT_CALLED_STATION_ON_CREATE', False
)
OPENVPN_DATETIME_FORMAT = get_settings_value(
    'OPENVPN_DATETIME_FORMAT', u'%a %b %d %H:%M:%S %Y'
)

TRAFFIC_COUNTER_CHECK_NAME = get_settings_value(
    'TRAFFIC_COUNTER_CHECK_NAME', 'Max-Daily-Session-Traffic'
)
TRAFFIC_COUNTER_REPLY_NAME = get_settings_value(
    'TRAFFIC_COUNTER_REPLY_NAME', 'CoovaChilli-Max-Total-Octets'
)
RADCLIENT_ATTRIBUTE_DICTIONARIES = get_settings_value(
    'RADCLIENT_ATTRIBUTE_DICTIONARIES', []
)

# counters
COUNTERS_POSTGRESQL = (
    'immunity_radius.counters.postgresql.daily_counter.DailyCounter',
    'immunity_radius.counters.postgresql.daily_traffic_counter.DailyTrafficCounter',
    'immunity_radius.counters.postgresql.monthly_traffic_counter.MonthlyTrafficCounter',
)
COUNTERS_MYSQL = (
    'immunity_radius.counters.mysql.daily_counter.DailyCounter',
    'immunity_radius.counters.mysql.daily_traffic_counter.DailyTrafficCounter',
    'immunity_radius.counters.mysql.monthly_traffic_counter.MonthlyTrafficCounter',
)
COUNTERS_SQLITE = (
    'immunity_radius.counters.sqlite.daily_counter.DailyCounter',
    'immunity_radius.counters.sqlite.daily_traffic_counter.DailyTrafficCounter',
    'immunity_radius.counters.sqlite.monthly_traffic_counter.MonthlyTrafficCounter',
)
DEFAULT_COUNTERS = {
    'django.db.backends.postgresql': COUNTERS_POSTGRESQL,
    'django.db.backends.mysql': COUNTERS_MYSQL,
    'django.db.backends.sqlite3': COUNTERS_SQLITE,
    # GIS backends
    'django.contrib.gis.db.backends.postgis': COUNTERS_POSTGRESQL,
    'django.contrib.gis.db.backends.mysql': COUNTERS_MYSQL,
    'django.contrib.gis.db.backends.spatialite': COUNTERS_SQLITE,
    'immunity_utils.db.backends.spatialite': COUNTERS_SQLITE,
}

try:
    _counters = get_settings_value(
        'COUNTERS', DEFAULT_COUNTERS[settings.DATABASES['default']['ENGINE']]
    )
except KeyError as e:  # pragma: no cover
    raise ImproperlyConfigured(str(e))

RADIUS_ATTRIBUTES_TYPE_MAP = get_settings_value('RADIUS_ATTRIBUTES_TYPE_MAP', {})

COUNTERS = []
CHECK_ATTRIBUTE_COUNTERS_MAP = {}
for counter_path in _counters:
    try:
        counter_class = import_string(counter_path)
    except ImportError as e:  # pragma: no cover
        raise ImproperlyConfigured(str(e))
    COUNTERS.append(counter_class)
    CHECK_ATTRIBUTE_COUNTERS_MAP[counter_class.check_name] = counter_class


# Extend the EXPORT_USERS_COMMAND_CONFIG[fields]
if not hasattr(settings, 'IMMUNITY
_USERS_EXPORT_USERS_COMMAND_CONFIG'):
    from immunity_users import settings as ow_users_settings

    ow_users_settings.EXPORT_USERS_COMMAND_CONFIG['fields'].extend(
        ['registered_user.method', 'registered_user.is_verified']
    )
    ow_users_settings.EXPORT_USERS_COMMAND_CONFIG['select_related'].extend(
        ['registered_user']
    )
